File: src/types/event.ts
export type EventPayload =
  | RawKeyEvent
  | MouseButtonEvent
  | MouseMoveEvent
  | MouseWheelEvent;

export interface RawKeyEvent {
  type: "KeyEvent";
  pressed: boolean;
  name: string;
}

export interface MouseButtonEvent {
  type: "MouseButtonEvent";
  pressed: boolean;
  button: MouseButton;
}

export interface MouseMoveEvent {
  type: "MouseMoveEvent";
  x: number;
  y: number;
}

export interface MouseWheelEvent {
  type: "MouseWheelEvent";
  delta_x: number;
  delta_y: number;
}

export type MouseButton =
  | "Left"
  | "Right"
  | "Middle"
  | "Back"
  | "Forward"
  | "Other";

export const RawKey = {
  // ───────────── Modifiers ─────────────
  ShiftLeft: "ShiftLeft",
  ShiftRight: "ShiftRight",
  ControlLeft: "ControlLeft",
  ControlRight: "ControlRight",
  Alt: "Alt",
  MetaLeft: "MetaLeft",
  MetaRight: "MetaRight",
  CapsLock: "CapsLock",
  Function: "Function",

  // ───────────── Navigation ─────────────
  UpArrow: "UpArrow",
  DownArrow: "DownArrow",
  LeftArrow: "LeftArrow",
  RightArrow: "RightArrow",
  Home: "Home",
  End: "End",
  PageUp: "PageUp",
  PageDown: "PageDown",
  Insert: "Insert",
  Delete: "Delete",

  // ───────────── Editing / Control ─────────────
  Return: "Return",
  KpReturn: "KpReturn",
  Tab: "Tab",
  Backspace: "Backspace",
  Escape: "Escape",
  Space: "Space",
  PrintScreen: "PrintScreen",
  ScrollLock: "ScrollLock",
  Pause: "Pause",
  NumLock: "NumLock",

  // ───────────── Function keys ─────────────
  F1: "F1",
  F2: "F2",
  F3: "F3",
  F4: "F4",
  F5: "F5",
  F6: "F6",
  F7: "F7",
  F8: "F8",
  F9: "F9",
  F10: "F10",
  F11: "F11",
  F12: "F12",
  // F13: "F13", 
  // F14: "F14", 
  // F15: "F15",
  // F16: "F16", 
  // F17: "F17", 
  // F18: "F18",
  // F19: "F19", 
  // F20: "F20", 
  // F21: "F21",
  // F22: "F22", 
  // F23: "F23", 
  // F24: "F24",

  // ───────────── Number row ─────────────
  Num1: "Num1",
  Num2: "Num2",
  Num3: "Num3",
  Num4: "Num4",
  Num5: "Num5",
  Num6: "Num6",
  Num7: "Num7",
  Num8: "Num8",
  Num9: "Num9",
  Num0: "Num0",

  // ───────────── Letters ─────────────
  KeyA: "KeyA",
  KeyB: "KeyB",
  KeyC: "KeyC",
  KeyD: "KeyD",
  KeyE: "KeyE",
  KeyF: "KeyF",
  KeyG: "KeyG",
  KeyH: "KeyH",
  KeyI: "KeyI",
  KeyJ: "KeyJ",
  KeyK: "KeyK",
  KeyL: "KeyL",
  KeyM: "KeyM",
  KeyN: "KeyN",
  KeyO: "KeyO",
  KeyP: "KeyP",
  KeyQ: "KeyQ",
  KeyR: "KeyR",
  KeyS: "KeyS",
  KeyT: "KeyT",
  KeyU: "KeyU",
  KeyV: "KeyV",
  KeyW: "KeyW",
  KeyX: "KeyX",
  KeyY: "KeyY",
  KeyZ: "KeyZ",

  // ───────────── Punctuation ─────────────
  BackQuote: "BackQuote",
  Minus: "Minus",
  Equal: "Equal",
  LeftBracket: "LeftBracket",
  RightBracket: "RightBracket",
  BackSlash: "BackSlash",
  SemiColon: "SemiColon",
  Quote: "Quote",
  Comma: "Comma",
  Dot: "Dot",
  Slash: "Slash",

  // ───────────── Numpad ─────────────
  Kp0: "Kp0",
  Kp1: "Kp1",
  Kp2: "Kp2",
  Kp3: "Kp3",
  Kp4: "Kp4",
  Kp5: "Kp5",
  Kp6: "Kp6",
  Kp7: "Kp7",
  Kp8: "Kp8",
  Kp9: "Kp9",
  KpPlus: "KpPlus",
  KpMinus: "KpMinus",
  KpMultiply: "KpMultiply",
  KpDivide: "KpDivide",
  KpDecimal: "KpDecimal",
  KpEqual: "KpEqual",
  KpComma: "KpComma",

  // ───────────── Media ─────────────
  VolumeUp: "VolumeUp",
  VolumeDown: "VolumeDown",
  VolumeMute: "VolumeMute",

  // ───────────── Mouse (Virtual) ─────────────
  Left: "Left",
  Middle: "Middle",
  Right: "Right",
  Back: "Back",
  Forward: "Forward",
  Drag: "Drag",
  ScrollUp: "ScrollUp",
  ScrollDown: "ScrollDown",
} as const;

export type RawKeyValue = typeof RawKey[keyof typeof RawKey];
export const MappedKeys = new Set<string>(Object.values(RawKey));

export const MODIFIERS = new Set<string>([
  RawKey.ShiftLeft,
  RawKey.ShiftRight,
  RawKey.ControlLeft,
  RawKey.ControlRight,
  RawKey.Alt,
  RawKey.MetaLeft,
  RawKey.MetaRight,
  RawKey.Function,
]);

export class KeyEvent {
  name: string;
  pressedCount: number;
  lastPressedAt: number;

  constructor(name: string) {
    this.name = name;
    this.pressedCount = 1;
    this.lastPressedAt = Date.now();
  }

  press() {
    this.pressedCount += 1;
    this.lastPressedAt = Date.now();
  }

  isModifier(): boolean {
    return MODIFIERS.has(this.name);
  }

  isNumpad(): boolean {
    return this.name.startsWith("Kp");
  }

  isArrow(): boolean {
    return this.name.endsWith("Arrow");
  }

  in(set: string[]): boolean {
    return set.includes(this.name);
  }
}

File: src-tauri/src/event.rs
use rdev::{Button, Event, EventType, Key};
use serde::Serialize;

#[derive(Debug, Clone, Serialize)]
#[serde(tag = "type")]
pub enum InputEvent {
    KeyEvent {
        pressed: bool,
        name: String,
    },
    MouseButtonEvent {
        pressed: bool,
        button: MouseButton,
    },
    MouseMoveEvent {
        x: f64,
        y: f64,
    },
    MouseWheelEvent {
        delta_x: i64,
        delta_y: i64,
    },
}

#[derive(Debug, Clone, Serialize)]
pub enum MouseButton {
    Left,
    Right,
    Middle,
    Other,
}

pub fn log_event(event: &Event) {
    match &event.event_type {
        EventType::KeyPress(key) => {
            println!("[KEY DOWN] key={:?}", key);
        }
        EventType::KeyRelease(key) => {
            println!("[KEY UP] key={:?}", key);
        }
        EventType::ButtonPress(button) => {
            println!("[MOUSE DOWN] button={:?}", button);
        }
        EventType::ButtonRelease(button) => {
            println!("[MOUSE UP] button={:?}", button);
        }
        EventType::MouseMove { x, y } => {
            println!("[MOUSE MOVE] x={}, y={}", x, y);
        }
        EventType::Wheel { delta_x, delta_y } => {
            println!("[MOUSE WHEEL] delta_x={}, delta_y={}", delta_x, delta_y);
        }
    }
}

pub fn convert_event(event: Event) -> Option<InputEvent> {
    match event.event_type {
        EventType::KeyPress(key) => {
            Some(InputEvent::KeyEvent {
                pressed: true,
                name: map_key(key),
            })
        }
        EventType::KeyRelease(key) => {
            Some(InputEvent::KeyEvent {
                pressed: false,
                name: map_key(key),
            })
        }
        EventType::ButtonPress(button) => {
            Some(InputEvent::MouseButtonEvent {
                pressed: true,
                button: map_mouse_button(button),
            })
        }
        EventType::ButtonRelease(button) => {
            Some(InputEvent::MouseButtonEvent {
                button: map_mouse_button(button),
                pressed: false,
            })
        }
        EventType::MouseMove { x, y } => {
            Some(InputEvent::MouseMoveEvent { x, y })
        }
        EventType::Wheel { delta_x, delta_y } => {
            Some(InputEvent::MouseWheelEvent { delta_x, delta_y })
        }
    }
}

pub fn map_key(key: Key) -> String {
    match key {
        // ───────────── Modifiers ─────────────
        Key::ShiftLeft | Key::ShiftRight => "Shift",
        Key::ControlLeft | Key::ControlRight => "Ctrl",
        Key::Alt | Key::AltGr => "Alt",
        Key::MetaLeft | Key::MetaRight => "Meta",
        Key::CapsLock => "Caps Lock",
        Key::Function => "Fn",

        // ───────────── Navigation ─────────────
        Key::UpArrow => "↑",
        Key::DownArrow => "↓",
        Key::LeftArrow => "←",
        Key::RightArrow => "→",
        Key::Home => "Home",
        Key::End => "End",
        Key::PageUp => "Page Up",
        Key::PageDown => "Page Down",
        Key::Insert => "Insert",
        Key::Delete => "Delete",

        // ───────────── Editing / Control ─────────────
        Key::Return => "Enter",
        Key::KpReturn => "Num Enter",
        Key::Tab => "Tab",
        Key::Backspace => "Backspace",
        Key::Escape => "Esc",
        Key::Space => "Space",
        Key::PrintScreen => "Print Screen",
        Key::ScrollLock => "Scroll Lock",
        Key::Pause => "Pause",
        Key::NumLock => "Num Lock",

        // ───────────── Function keys ─────────────
        Key::F1 => "F1",   Key::F2 => "F2",   Key::F3 => "F3",
        Key::F4 => "F4",   Key::F5 => "F5",   Key::F6 => "F6",
        Key::F7 => "F7",   Key::F8 => "F8",   Key::F9 => "F9",
        Key::F10 => "F10", Key::F11 => "F11", Key::F12 => "F12",
        // Key::F13 => "F13", Key::F14 => "F14", Key::F15 => "F15",
        // Key::F16 => "F16", Key::F17 => "F17", Key::F18 => "F18",
        // Key::F19 => "F19", Key::F20 => "F20", Key::F21 => "F21",
        // Key::F22 => "F22", Key::F23 => "F23", Key::F24 => "F24",

        // ───────────── Number row ─────────────
        Key::Num1 => "1", Key::Num2 => "2", Key::Num3 => "3",
        Key::Num4 => "4", Key::Num5 => "5", Key::Num6 => "6",
        Key::Num7 => "7", Key::Num8 => "8", Key::Num9 => "9",
        Key::Num0 => "0",

        // ───────────── Letters ─────────────
        Key::KeyA => "A", Key::KeyB => "B", Key::KeyC => "C",
        Key::KeyD => "D", Key::KeyE => "E", Key::KeyF => "F",
        Key::KeyG => "G", Key::KeyH => "H", Key::KeyI => "I",
        Key::KeyJ => "J", Key::KeyK => "K", Key::KeyL => "L",
        Key::KeyM => "M", Key::KeyN => "N", Key::KeyO => "O",
        Key::KeyP => "P", Key::KeyQ => "Q", Key::KeyR => "R",
        Key::KeyS => "S", Key::KeyT => "T", Key::KeyU => "U",
        Key::KeyV => "V", Key::KeyW => "W", Key::KeyX => "X",
        Key::KeyY => "Y", Key::KeyZ => "Z",

        // ───────────── Punctuation ─────────────
        Key::BackQuote => "`",
        Key::Minus => "-",
        Key::Equal => "=",
        Key::LeftBracket => "[",
        Key::RightBracket => "]",
        Key::BackSlash | Key::IntlBackslash => "\\",
        Key::SemiColon => ";",
        Key::Quote => "'",
        Key::Comma => ",",
        Key::Dot => ".",
        Key::Slash => "/",

        // ───────────── Numpad ─────────────
        Key::Kp0 => "Num 0", Key::Kp1 => "Num 1", Key::Kp2 => "Num 2",
        Key::Kp3 => "Num 3", Key::Kp4 => "Num 4", Key::Kp5 => "Num 5",
        Key::Kp6 => "Num 6", Key::Kp7 => "Num 7", Key::Kp8 => "Num 8",
        Key::Kp9 => "Num 9",
        Key::KpPlus => "Num +",
        Key::KpMinus => "Num -",
        Key::KpMultiply => "Num *",
        Key::KpDivide => "Num /",
        Key::KpDecimal => "Num .",
        Key::KpEqual => "Num =",
        Key::KpComma => "Num ,",

        // ───────────── Media ─────────────
        Key::VolumeUp => "Volume +",
        Key::VolumeDown => "Volume -",
        Key::VolumeMute => "Mute",

        // ───────────── Language / IME ─────────────
        // Key::Lang1 => "Lang 1",
        // Key::Lang2 => "Lang 2",
        // Key::Lang3 => "Lang 3",
        // Key::Lang4 => "Lang 4",
        // Key::Lang5 => "Lang 5",
        // Key::Kana | Key::KanaMode => "Kana",
        // Key::Hangul => "Hangul",
        // Key::Hanja | Key::Hanji => "Hanja",
        // Key::Junja => "Junja",
        // Key::Final => "Final",

        // ───────────── System / Legacy ─────────────
        Key::Apps => "Menu",
        Key::Help => "Help",
        Key::Sleep => "Sleep",
        Key::Select => "Select",
        Key::Execute => "Execute",
        Key::Print => "Print",
        Key::Clear => "Clear",
        Key::Cancel => "Cancel",
        Key::Separator => "Separator",

        // ───────────── Unknown / Raw ─────────────
        // Key::RawKey(_) => "RawKey",
        _ => "Unknown",
    }
    .to_string()
}

pub fn map_mouse_button(button: Button) -> MouseButton {
    match button {
        Button::Left => MouseButton::Left,
        Button::Right => MouseButton::Right,
        Button::Middle => MouseButton::Middle,
        _ => MouseButton::Other,
    }
}